### ISSUE TITLE
Wanted: Experience Reports

### ISSUE THREAD

**Issue by joelhawksley** [github/view_component#15]
👋 thanks for checking out `actionview-component`!

In sharing this gem with the Rails community, we are looking to hear about your experiences with using `ActionView::Component` in your applications.

Please let us know what works well, what doesn't, and what you think is missing ❤️ 

If you'd rather provide feedback privately, feel free to contact me at joelhawksley@github.com.

**Comment by metade** [comment]
We've started using `ActionView::Component` in our application and it's really helped clean up some nasty views/helper/presenter combos.

Thank you so much for kickstarting this - it feels to me that `ActionView::Component` was an important missing piece of the Rails toolset and it's great to see this come together.

## Issues 

We've noticed a few issues in our application (Rails 5.2) that I'll start by listing here.

We've found some workarounds but I'm not sure if they are the best way to fix them - I'm happy to create a PR/issue if you think they have potential!

### Unable to render forms/button_to from a component

Rendering a form/button_to from a component raises an error `undefined method 'protect_against_forgery?' for #<Happening::SignUpButton:0x00007f893937e7f8>`

### Unable to render partials and use `content_for` from a component

It was useful when refactoring existing partials/helpers to reuse existing partials/context_for.

Rendering partials caused a `undefined method `render_partial' for nil:NilClass` due to the view renderer not being set.

Using `content_for` caused `undefined method `append' for nil:NilClass` due to the `view_flow` not being set.

## Workarounds

We've created an `ApplicationComponent` for our app to fix these:

```
class ApplicationComponent < ActionView::Component
  include ActiveSupport::Configurable
  include ActionController::RequestForgeryProtection

  def render_in(view_context, *args, &block)
    # Enable button_to/forms in components
    self.controller = view_context.controller
    # Enable partials to be rendered in components
    @view_renderer ||= view_context.view_renderer
    # Enable content_for to be used in components
    @view_flow ||= view_context.view_flow

    super
  end
end
```

We also had to override the test helper as follows:

```
module ActionView
  module ComponentTestHelpers
    def render_component(component, &block)
      controller = ApplicationController.new
      controller.request = ActionDispatch::TestRequest.create
      Nokogiri::HTML(component.render_in(controller.view_context, &block))
    end
  end
end
```

As I said above, happy to kickstart an issue/PR if you feel that any of these workarounds have potential.

## Other issues

In development, we've noticed that when a component view is changed Rails doesn't pick up the change unless the component object file is also updated. 

It's annoying but I've not been annoyed enough to try and find a workaround/fix!

Finally, we were had a component without a view `.template`, I noticed when I switched to the gem that you've removed that ability but the error message still refers to the `.template` approach:
> "Could not find a template for #{self}. Either define a .template method or add a sidecar template file."

**Comment by mikker** [comment]
First; I love this so much. Had the exact same experience of coming back from React and "missing" the component way to structure things. And testing – this is so much better ❤️

I'll second the experiences of @metade – in development, changes aren't always picked up. Perhaps it could be solved as easy as disabling template compilation in development (which I think is happening?)

**Comment by vinistock** [comment]
cc @jennaleeb

We added the gem to a Rails 6.0.0.beta3 project and ran into an issue when rendering a test component (both in dev and test).

Inside the `DummyTemplate` class, `identifier` is hard coded to an empty string, but better_html tries to [split this value](https://github.com/Shopify/better-html/blob/master/lib/better_html/better_erb.rb#L43), which leads to a method being invoked on nil.

The irony is not lost on us that better_html is a Shopify project. We're wondering if someone else ran into the issue and if we're suppose to set the identifier somehow.

**Comment by joelhawksley** [comment]
@metade @mikker thank you SO much for your experience reports. This is _exactly_ the kind of feedback we are looking for.

The local reloading issue definitely needs to be addressed. I've filed an issue here: https://github.com/github/actionview-component/issues/19 and have started working on a fix.

@metade generally, it sounds like we need more of the context Rails provides views when being rendered. I'd prefer to come up with a solution that allows for components to not be coupled to the request, allowing them to be used outside of the request path (such as in jobs or mailers). However, we'll probably need to at least inject a dummy controller for things like rendering forms.

This is an issue that has come up internally that we haven't really dug into yet, but plan to look at soon. For now, your solution of injecting the controller and view flow context seems directionally correct.

**Comment by inopinatus** [comment]
Took AV::C for a test drive today. Here's some freeform impressions.

I immediately started rearranging the innards. The naming conventions blew some fuses. The first was renaming `app/components` to `app/views/components`. AV::C is an element of ActionView, not a whole new subsystem. The revised path is less presumptuous (and avoids conflict), it's also more intention-revealing.  Every component class had `Component` appended. This happens for jobs, mailers, controllers, helpers, channels; only _model_ objects get a pass on having their role appended.  Similarly, `ActionView::Component`becomes `ActionView::Component::Base` so I can `class ApplicationComponent < AV::C::B` in keeping with every other Rails subsystem.

The testing story is really nice.  Every large team & large application is crying out for testable view objects with programmatic quality control.  This is AV::C's biggest win.

The collection rendering story isn't great. Collection partials were 25x faster over component iteration in my testing.  However, I can't see a way to achieve a flyweight render function without dumping validations.

My biggest concern: there's little-to-no convention-over-configuration going on.  One of the stated benefits of AV::C is in reducing conditionals, and yet there's conditionals in the demo because only explicit class selection is supported.  In the end I still really want to write `render @model` but have Rails try `#to_component_class` before `#to_partial_path`. And I still want namespacing and all the other jazz.

To me, right now, it feels like a viable OO replacement for helpers and for value objects, but not (yet) for rendering entity models and collections. I'm not clear on whether that was the intention or not. More broadly I've got a concern that AV::C is not a terribly enthusiastic participant in the Rails framework.

**Comment by jensljungblad** [comment]
@topherfangio Thanks for checking out https://github.com/jensljungblad/components. I always thought with Ruby we should be able to one-up React :) If `AV:C` won't incorporate a DSL, perhaps a DSL could build on top of `AV:C`.

@joelhawksley I was checking out the sample application and I've got a question!

```erb
<%= render(Issues::Badge, state: issue.state.to_sym) %>
```

This passes a `state` symbol to the component. But is it possible to use this data to create new methods, override it, etc, in the component class? In the example, most of the logic exists in the template:

```erb
<%= render(Primer::State, color: STATES[state][:color], title: "Status: #{state.to_s.titleize}") do %>
  <%= octicon(STATES[state][:octicon_name]) %> <%= STATES[state][:label] %>
<% end %>
```

Why not have methods on the component that fetches the `color`, the `label`, the `octicon_name` and does the titleize? I would assume that is one point of creating Ruby classes in the first place, as a place to put logic. In the example, the only thing the component class does with the passed `state` is to validate its presence. I guess I would assume something like `render component with data -> component initialized -> component transforms data -> template receives transformed data`.

**Comment by topherfangio** [comment]
@jensljungblad So, that should be totally possible. When you create the component, you can define any methods you wish, and they can definitely be based off of any instance variables.

So something like this should be totally valid:

```ruby
class Issues::BadgeComponent < ApplicationComponent

  COLORS = [open: :green, merged: :purple, closed: :red]

  validates :state, presence: true

  def initialize(state:)
    @state = state
  end

  def color
    COLORS[@state]
  end

  private

  attr_reader :state
end
```

**Comment by jensljungblad** [comment]
@topherfangio How would you access them from the template?

**Comment by mikker** [comment]
@jensljungblad they're available directly, like `<%= color %>` would output the result of the `color` method.

**Comment by jensljungblad** [comment]
Strange, I got errors when I tried that. I'll try again :)

**Comment by topherfangio** [comment]
@joelhawksley @jensljungblad 

I have created a gist that shows how to get AV:C working with the DSL provided by https://github.com/jensljungblad/components!

I have some notes/thoughts at the end about some issues I ran into while building this. The biggest one is that debugging is very difficult because it seems to die on the `render(MyComponent)` call inside the HAML file instead of showing me the actual line of code in the component that is failing.

https://gist.github.com/topherfangio/76615f32f7ebb25e3caa03741cb13a59

It's *usuable* as-is, but I feel like there is a lot of room for improvement.

Would love y'all's feedback! Thanks!

**Comment by cjbottaro** [comment]
It doesn't work with:
- forms
- other partials
- content_for

We went down that path of storing the view context and setting certain instance variables so that various `ActionView::Base` methods would work, but ultimately gave up. It just felt dirty and hack-y, and I'm guessing it won't work anyway because of Rail's rendering buffers.

**Comment by cjbottaro** [comment]
https://github.com/github/actionview-component/pull/28 with a little tweak to get it working with `content_for` addressed all my issues above.

So now after actually using it...

It causes a proliferation of classes and files in my codebase. Consider the example of a PostsController, and the feasible components that go along with basic crud: index, index_item, new, edit, show, form. That's 10 files and 5 new classes. 4 out of 5 of the classes are identical except for the names.

I would rather it be like controllers, i.e. "a controller has many actions", ala "a component has many templates". This is how Cells and Phoenix work as well, and I think it cuts down files/classes/module and makes code reuse/sharing easier. That means validation will have to support `:only` and `:except`.

Fwiw, I think validations declared on the class is an antipattern (I think this about ActiveRecord also).

```ruby
class PostComponent < ApplicationComponent

  attr_reader :post, :posts

  def initialize(post: nil, posts: nil)
    @post = post
    @posts = posts
  end

  def index
    validates_presence_of(:posts)
    render :index
  end

  def show
    validates_presence_of(:post)
    render :post
  end

end
```

Using this more procedural approach, validation can be conditional using simple if statements, and grouped together for reuse via functions (methods).

Otherwise, implementing conditional validation on the class becomes needlessly complex:
```ruby
class PostComponent < ApplicationComponent
  validates_presence_of :post, if: :some_method
  validates_presence_of :post, if: ->{ blah }
  validates_presence_of :post, only: [:new, :show, :edit]
end
```

**Comment by inopinatus** [comment]
> Consider the example of a PostsController, and the feasible components that go along with basic crud: index, index_item, new, edit, show, form. 

I'm not sure that's the intended pattern. This seems congruent to rendering views according to controller actions, just with more scaffolding, and it sounds like you're looking at it and wondering "what's the point?", which is reasonable given the outcome.

_warning: strongly held opinions from this point on_

I have some thoughts on how AV:C "ought to be used". By which I mean, how it can fill a niche that isn't already something Rails does very well indeed.

Here's the thing; I think we already have a very good framework for rendering based on polymorphic models, routing, and controller actions: Action Pack; and Action Pack is powerful and capable already and not, I reckon, a wheel in urgent need of re-invention.

When spiking AV:C I followed the apparent design intention of designing each component for reusability, taking cues from the [README](https://github.com/github/actionview-component#when-should-i-use-components) and from the [demo repository](https://github.com/joelhawksley/actionview-component-demo) and the [RailsConf talk](https://www.youtube.com/watch?v=y5Z5a6QdA-M). One component per (model, action) doesn't fit that narrative, and I didn't need, want, or create a separate component for each such tuple.

### What components are

To me, components don't represent models or actions; they exist as abstractions of _page-level objects_.

That's informed by using reactive client-side rendering frameworks, particularly Vue.js (but also looking longingly at the charming simplicity of Svelte), and having a great experience therein, and then wishing for the same but for server-rendered HTML.

When kicking the tyres on AV:C, my components were:

* `CardComponent` that can accept any model responding to `title`, `image`, `text` and generate appropriate card HTML.  
* `FormComponent` that is essentially form_for turned inside out and wrapped in an object along with form building preferences.  
* `Input::BaseComponent` abstract base for `Input::StringComponent`, `Input::CheckboxComponent`, `Input::DobComponent` styled and wrapped how I want them.
* `TableComponent` that renders any collection with automatic headings.

To me, AV:C is what helpers are when they grow up and become object-oriented.

I could foresee creating a

* `NavComponent` for my navbar, and various sub-components e.g. `LogoComponent`, `AvatarComponent`.
* Obvious widgets like a `DropdownComponent` or a `TagListComponent`
* `Input::DropdownComponent` that uses the above to select from an association collection
* `LoginComponent` that is our sign-in form.
* Abstract page structures like `ModalComponent`, `OpenGraphComponent`, `GoogleAnalyticsComponent`.
* Component(s) for in-place editing, and anything that `select2`-style input libraries do.
* `AutoFormComponent` which takes any model object and makes a form of its fields, like some kind of magical instant scaffolding. Configuration optional.
* `TableComponent` growing to support the injection of abstract pagination & search facilities, automatic columns for edit/delete buttons, an inline append form.

Between those last two I imagined a new line of admin interface toolkit emerging.

I'd also want to find a clear and elegant story bringing AV:C and Stimulus together.

### Validation

> Fwiw, I think validations declared on the class is an antipattern

I agree with this in principle, but mainly because I think component validations are likely unnecessary in most cases and represent a severe performance drag for collections.

I would very much like to be able to write a component that is basically a sophisticated closure, and can be instantiated once for an entire collection. Much like `render partial: @collection` does. Validation callbacks (or callbacks of any kind) effectively preclude that.

Much like type systems, and particularly being mindful that some kind of type system is likely in future versions of Ruby, I think validation affords the greatest programmer happiness when it is optional, and incremental, but still there when you need it. 

This is one way a large team can benefit from formal methods whilst co-existing in the same ecosystem as swashbuckling solo MVP hackers.

Maybe make validations an optional `include ActionView::Component::Validation`? I'm looking at the `Base` class and thinking it could probably be carved up into composable concerns.

### End goal

If I could write in a controller

    render IndexPageComponent, collection: @whatever

and everything just sort of unrolls from there, I'd be a very happy dev.

**Comment by czj** [comment]
We've been using `ActionView::Component` with great pleasure to cleanup helpers, presenters, decorators and some tiny partials. Using and testing are a real pleasure.

I would suggest adding to the documentation a more prominent section about having a component that renders directly without a template, for people how have complex logic but need, at the end, only a `content_tag` with a large hash of options.

**Comment by cjbottaro** [comment]
That makes a lot of sense. So in my example, `index, show, new, edit` would remain normal Rails views and `index_item` (arguable) and `form` would be good candidates for AV:C.

I think a lot of my views (pun) are skewed by my love for Phoenix, so seeing render calls that look almost identical started to give me tunnel vision.

**Comment by eshnil** [comment]
> Fwiw, I think validations declared on the class is an antipattern

@cjbottaro @inopinatus I think validations come in two flavors. (1) To check integrity of the record (2) Business logic. The first kind is best declared at the class. And the second one should be pure function calls the way you showed. In fact, if ActiveRecord integrate better with the underlying databases, (1) would be directly mapped to the `CHECK` constraints on the table.

Conditional validations at class-level, for both ActiveRecord and ActionView::Component are indicative of business logic being put in place of integrity checks.

**Comment by kinduff** [comment]
Good experience so far, sometimes component creation (specially attr_reader) feels repetitive but I like it. 

One of the pain issues I've had is testing with Rspec using helpers (based on #105) and Devise. Had to dirty fix my tests with `type: :controller` to make them work, besides adding `config.include ActionView::Component::TestHelpers, type: :controller` to my `rails_helper.rb`.

I'm going to experiment on the next few days with being able to create a gem with components that I would like to reuse in the future and also to create a styleguide.

**Comment by juanmanuelramallo** [comment]
@kinduff Check #120 
Would that help?

**Comment by kinduff** [comment]
@juanmanuelramallo looks like I'm already doing exactly that, happy to see there's a PR for it. I switched to treat components as controllers with Rspec and I'm also including the `Component::TestHelpers` class. I'll post my findings in the PR just in case.

**Comment by bostanio** [comment]
I've been running some simple experiments with AV:C and like it a great deal. The testability is key and a bit reason I suspect we will adopt this widely. A couple of thoughts/comments:

- One folder per component seems key for large codebases #67. 
  It sounds like GitHub has experience with a bunch of components and a larger codebase and found organizing by modules to be enough. I struggle to see how this pattern plays out well at scale.
- A solution for templating multiple rendering outlets as discussed in #144.
- Documentation and best practices for including js, css and other assets with a component.
  Deep automated integration into Webpack(er) would be nice for this too. Mirroring the common React pattern of putting all the supporting assets in the component's folder would be fantastic. I can see components coming with their own css and simple JS enhancements via Stimulus controllers. Making that pattern slick would be very powerful.
  I'd be particularly curious if we can eventually end up with "scoped CSS" kinds of patterns inspired by React projects like [styled-components](https://github.com/styled-components/styled-components)
- Documentation, best practices and examples on where ViewComponents should sit in the overall MVC pattern. 
  I appreciate that part of the power of Rails is the ability to mould it to your own use. However, in general there are a bunch common patterns on how to package data for views. Folks are using combinations of non-AR Models, Facades, Presenters, Decorators etc. Where does AV:C sit? What should AV:Components be responsible for and when should you use something else?
- Make Preview as slick as Storybook.JS!
  I'm sure there is a bunch of work here but but getting close to the slickness of Storybook would really help building out a Design System documented in the browser.

**Comment by joelhawksley** [comment]
@bostanio thank you for the thoughtful feedback ❤️ 

> * One folder per component seems key for large codebases #67.
>   It sounds like GitHub has experience with a bunch of components and a larger codebase and found organizing by modules to be enough. I struggle to see how this pattern plays out well at scale.

This remains the case for us internally. @mellowfish is using a fair number of components, perhaps he could provide examples of his folder/module structure?

> * A solution for templating multiple rendering outlets as discussed in #144.

I agree that this would be _awesome_. I don't think we need to have this before `v2`, but it's a killer idea that would give us something _better_ than what React offers, in my opinion.

> * Documentation and best practices for including js, css and other assets with a component.
>   Deep automated integration into Webpack(er) would be nice for this too. Mirroring the common React pattern of putting all the supporting assets in the component's folder would be fantastic. I can see components coming with their own css and simple JS enhancements via Stimulus controllers. Making that pattern slick would be very powerful.
>   I'd be particularly curious if we can eventually end up with "scoped CSS" kinds of patterns inspired by React projects like [styled-components](https://github.com/styled-components/styled-components)

I'd be _very_ interested in seeing a proof of concept for this. I'm wary of adding any functionality in this direction _unless_ there is automatic scoping of any sibling CSS/JS. Otherwise we risk just introducing confusion by implying encapsulation that doesn't exist.

> * Documentation, best practices and examples on where ViewComponents should sit in the overall MVC pattern.
>   I appreciate that part of the power of Rails is the ability to mould it to your own use. However, in general there are a bunch common patterns on how to package data for views. Folks are using combinations of non-AR Models, Facades, Presenters, Decorators etc. Where does AV:C sit? What should AV:Components be responsible for and when should you use something else?

This is literally the topic of the next conference talk I'm starting to write on this project.

My current thinking is this: ActionView::Components are a _superset_ of templates. Rails has essentially hidden the Components abstraction up until now.

I'm working on an explainer of the existing Rails template rendering architecture over the next couple of weeks, with the hope that it will help us better communicate how components fit into the Rails MVC pattern.

I'd love to hear how others see components fit into their Rails mental model.

> * Make Preview as slick as Storybook.JS!
>   I'm sure there is a bunch of work here but but getting close to the slickness of Storybook would really help building out a Design System documented in the browser.

I'm not super familiar with Storybook. Might you be willing to open an issue describing what kind of improvements you envision us making to the existing implementation of previews?

**Comment by mellowfish** [comment]
@bostanio I have been organizing my components like I would models or anything else. Where it makes sense, I have module namespaces (sometimes deep ones), and appropriate folders. 

But I am not sure I would agree with one folder per component, that seems excessive. Maybe if there were a way to optionally handle that (same folder name as component, without expecting a module namespace). If we start getting assets like JS and CSS I could see where it could be useful, but for simple view/component pairs, I like being able to have several related ones together in a folder.

**Comment by bostanio** [comment]
> @bostanio I have been organizing my components like I would models or anything else. Where it makes sense, I have module namespaces (sometimes deep ones), and appropriate folders.
> 
> But I am not sure I would agree with one folder per component, that seems excessive. Maybe if there were a way to optionally handle that (same folder name as component, without expecting a module namespace). If we start getting assets like JS and CSS I could see where it could be useful, but for simple view/component pairs, I like being able to have several related ones together in a folder.

@mellowfish I completely here you on view/component pairs. My commentary is more about including assets JS, CSS and Images (SVGs) 'scoped' to a component. Of course the challenge is that Ruby has very particular expectations around folders meaning namespaces. I certainly don't want a namespace for every component that would be overkill. Maybe due to the constraints of Ruby we can't offer other options but it would be worth exploring.

**Comment by bostanio** [comment]
> > * A solution for templating multiple rendering outlets as discussed in #144.
> 
> I agree that this would be _awesome_. I don't think we need to have this before `v2`, but it's a killer idea that would give us something _better_ than what React offers, in my opinion.

Agreed it doesn't need to before v2 but it would be slick.

> > * Documentation and best practices for including js, css and other assets with a component.
> >   Deep automated integration into Webpack(er) would be nice for this too. Mirroring the common React pattern of putting all the supporting assets in the component's folder would be fantastic. I can see components coming with their own css and simple JS enhancements via Stimulus controllers. Making that pattern slick would be very powerful.
> >   I'd be particularly curious if we can eventually end up with "scoped CSS" kinds of patterns inspired by React projects like [styled-components](https://github.com/styled-components/styled-components)
> 
> I'd be _very_ interested in seeing a proof of concept for this. I'm wary of adding any functionality in this direction _unless_ there is automatic scoping of any sibling CSS/JS. Otherwise we risk just introducing confusion by implying encapsulation that doesn't exist.

I agree scoping of CSS/JS would be a nice option but IMO its not a requirement. In React land there are solutions for scoping but you don't have to use them (and I've seen very successful, large code bases NOT use those solutions and simple rely on BEM naming conventions and avoid css class collisions). There are other advantages of organizing the files together even if you don't scope the CSS/JS:

1. It just makes it easy to find things (and easy to move/delete "all the things")
1. It indicates which assets (css definitions, Stimulus controllers, images etc.) are 'owned' by that component. You can cheat and share them in other places but its easy to spot (and I imagine possible to write linting rules to prevent)
1. Provides options for webpack bundling of a subset of components by glob including certain folders/namespaces (for example components only used by the Admin part of my site that I want to have separate JS/CSS bundles separate from the prod bundles)

You can get decent solutions to all of the above by enforcing parallel directory structures in app, javascript, spec folders etc. That pattern seemed entirely natural to me for a long time, it's very much "the Rails way" but having had folders per component in React it feels very powerful.

**Comment by bostanio** [comment]
@joelhawksley here is a Storyboook inspired ticket around Previews #159

**Comment by dylnclrk** [comment]
@topherfangio Brought up something that I've also found confusing. When creating a component that takes no arguments besides content, calling render raises "wrong number of arguments (given 1, expected 0)".

```
class MyComponent
   def initialize
   end
end
```

**Comment by ozzyaaron** [comment]
I'm just building on some previous reports primarily around where templates live and the filenames of the template/s. My concerns come down to integrating with Rails caching basics.  

The issue that we run into is basically summed up by 

```
## In any view/partial
cache[object, another_object] do
  render(MyComponent, attribute: value)
end
```

In this case if the template for `MyComponent` is changed the cache will not be invalidated. I tried to add `MyComponent` to the array used by `cache` and provide a `self.cache_key` method that used `ActionView::Digestor` but without having some further mechanism to invalidate the digestor's cache for the template/s when the template changes this didn't work. 

Rails allows for explicit directives to be used (https://guides.rubyonrails.org/caching_with_rails.html#explicit-dependencies) but by default this will only look inside `app/views`. You can add a view path `append_view_path("app/components")` but the digestor is expecting a partial and so a template filename not prefixed with underscore is not found. 

The way we are currently working around this is to have an `ApplicationComponent` that our components inherit from and where we override `ActionView::Component::Base.matching_views_in_source_location` so that templates are now found in `app/views/component_templates` and where the files are prefixed with underscore. 

```
  def self.matching_views_in_source_location
    return [] unless source_location

    *path, file_name = source_location.chomp(File.extname(source_location)).split("/")
    path.pop # remove components
    path.concat(["views", "component_templates"])
    path_to_template = path.push("_#{ file_name }").join("/")

    Dir["#{ path_to_template }.*{#{ActionView::Template.template_handler_extensions.join(',')}}"]
  end
```

Then in the above example where cache invalidation does not work via template digest changes we can do 

```
## In any view/partial
# Template Dependency: component_templates/my_component %>
cache[object, another_object] do
  render(MyComponent, attribute: value)
end
```

I think if AV::C is going to be core or integrate with Rails it would need a way to expose template digests nicely, or provide `cache_key` probably at the class level. If not then as a last resort we would need a way to use explicit directives in templates that use ActionView::Components internally. 

I'd also be all 👂 to people who have solved this already or if I've missed something about the features `ActionView::Component` provides!

**Comment by joelhawksley** [comment]
@ozzyaaron thank you for the detailed report!

Might you be willing to file this as an issue so we can track and discuss it further?

**Comment by ozzyaaron** [comment]
No problems @joelhawksley I have filed it! https://github.com/github/actionview-component/issues/234

**Comment by nicolas-brousse** [comment]
Hey! I currently work with `komponent` gem, and there is a while that I wanted to try `ActionView::Component`.

There is also some months that I was looking to do some tries with Rails form helper + components. So it was a good opportunity to try `ActionView::Component` too.

I did a small Rails project to test what I had in head. So my idea was simply to supercharge Rails form helper with components. And with `ActionView::Component` I was able to do it without big difficulties.

I have to dig a bit more into it, but I think it is something we could adopt in a close future in our processes.

Here the repository with a demo ➡️ https://github.com/nicolas-brousse/avc_form_builder#readme

I'm interested for feedback if any 🙂

**Comment by joshleblanc** [comment]
Is there any solution available for component scoped CSS?

Whether that's some other dependency, or built in?

**Comment by joelhawksley** [comment]
@joshleblanc not to my knowledge, no. It's on our radar, but not until later this year at the earliest.

**Comment by rosendi** [comment]
@joelhawksley Thank you for collection rendering! It's awesome!

**Comment by joelhawksley** [comment]
@rosendi don't thank me, thank @tclem! (and @jonspalmer / @rainerborene 😄 )

**Comment by tkoenig** [comment]
@joshleblanc @joelhawksley what I am currently trying in combination with stimulus, is scoping on the `data-controller[...]` in the css.

So in the `description_component_controller.js` file I import the css

```js 
import './description_component.scss'
```

and the css looks something like this:

```css
[data-controller="description-component"] {
  background-color: greenyellow;
}
```

If you don't need any javascript for the component, just using the import statement is fine as long you use webpack for css.

**Comment by abitrolly** [comment]
@joelhawksley I am a n00b in Ruby on Rails and found this while looking for how can I add Jekyll-style rendering of .md files to [this website](https://github.com/minsk-hackerspace/hackerspace.by). Over initial scanning of the text, I have no idea what a component is, other that it should probably meant to be used in `render` function. The thing that attracted me is the promise that I can unit test it. It it is possible without bootstrapping the whole Rails framework, that is awesome.

I would really appreciate a n00bs tutorial for people who just started with Rails and followed its tutorial to build their first `hello rails`. Why routing is not required, is there common config for all components, how are they subclassed, where I can use them from, can I use them without Rails at all, and most importantly - how to render .md blog post that is stored in GitHub repository in native Rails layout?

**Comment by pinzonjulian** [comment]
> @joelhawksley I am a n00b in Ruby on Rails and found this while looking for how can I add Jekyll-style rendering of .md files to [this website](https://github.com/minsk-hackerspace/hackerspace.by). Over initial scanning of the text, I have no idea what a component is, other that it should probably meant to be used in `render` function. The thing that attracted me is the promise that I can unit test it. It it is possible without bootstrapping the whole Rails framework, that is awesome.
> 
> I would really appreciate a n00bs tutorial for people who just started with Rails and followed its tutorial to build their first `hello rails`. Why routing is not required, is there common config for all components, how are they subclassed, where I can use them from, can I use them without Rails at all, and most importantly - how to render .md blog post that is stored in GitHub repository in native Rails layout?

Hey @abitrolly !
Good to see you're starting out with Rails! 

### On using it outside of Rails
I believe it's not possible at the moment (and probably not in the future). The objective of this gem is meant to work as a first class citizen in Rails and it depends on some of Rails' core modules. See [the docs](https://github.com/github/view_component#isnt-this-just-like-x-library) for a list of other Ruby libraries that probably don't depend on Rails for this.

### What's a component?
You can check the docs [here](https://github.com/github/view_component#what-are-components). In short: View component is a way to have a ruby object handle all the logic for rendering a piece of html just like JS handles all the logic to render html in React. Different from react, you have 2 files instead of 1: your template and the ruby file. All the logic you would add in your view would now be handled by the ruby object. 

### Starting out with Rails?
View-component is at the bleeding edge of what you can do in terms of rendering html to a page following Rails' conventions as close as possible so there aren't going to be many (if any) examples on the web right now to help you other that what you can find here. Maybe you can check out other, more mature implementations [here](https://github.com/github/view_component#isnt-this-just-like-x-library) that offer similar features and probably don't depend on Rails.

### Taking a shot at your questions
#### On routing: 
I'm guessing you're comparing View-component to React and how it uses routes to render components. This uses a totally different architecture. You navigate to a page with a regular http visit and then, inside the response, you render one or multiple components in your view. For example, if you have a `PostsController` with an `show` action that renders an `show.html.erb` template, you may write inside your template `= render PostComponent.new(title: @post.title, content: @post.content` to render the post.

#### Common config: 
Not sure what you mean by this. Most stuff in vanilla Rails don't need much configuration. Everything works out of the box and this gem is supposed to be that way too.

#### How are they subclassed?
You can create classes that inherit from `ApplicationComponent` to have common behavior between components if that's useful to you. Check out the docs [here](https://github.com/github/view_component#conventions)

#### On rendering markdown
Checkout redcarpet [here](https://github.com/vmg/redcarpet)

Hope this helps!
PS: If you need help getting started with rails, I strongly recommend the [Rails Tutorial](https://www.railstutorial.org/) book by Michael Hartl.

**Comment by joelhawksley** [comment]
@abitrolly welcome! @pinzonjulian has done a better job responding that I ever could! Thank you Julián!

**Comment by abitrolly** [comment]
@pinzonjulian thanks! Help to set the limits. I have little experience in JS, so don't know what React is. From what I know, React looks similar to game loop, where each frame is a iteration - data preparation, rendering, handling of input events. But maybe that's Flux. Not sure.

Among the Ruby on Rails tutorial I found this one to be the best - http://installfest.railsbridge.org/intro-to-rails/intro-to-rails - short practical steps with minimal text.

By `routing` I meant Rails `config/routes.rb`. Now I understand that I can not use the component directly, and still need the controller to and the view. Then in the view I need to reference my component in `render`. I guess all components from `app/components` are automatically available from any view, and I don't have to do additional wiring.

Summarizing. I can not build an isolated "component" that just renders markdown to refer to it directly from `config/routes.rb`, but I can create route, controller, view and then use the component to read referenced markdown file and render it if it is present.

**Comment by joshleblanc** [comment]
I wonder if there's a way to integrate StimulusReflex into view component. 

https://docs.stimulusreflex.com/quickstart

If your view component was a stimulus reflex controller itself, then it could obscure the line between front and backend.

**Comment by bbugh** [comment]
Firstly, hi 👋, and this is awesome 🔥. I really appreciate the extra effort y'all are doing to make this have third-party (and hopefully first-party) support in Rails. Rails lack of view objects for partials has made me crazy since it came out, and this is the best solution I've seen yet.

The below is written from the perspective of a team that already has a large Vue-based application (which we are keeping), but also needs to add some other complex reactive features to the server-rendered part of the site (like for admin and customer billing, etc.).

As far as I can tell, ViewComponent supports most of our use case, but there's a few major missing features for our large app needs.

**Multiple views per component**

Allowing ViewComponent to use one object (`PricingSummaryComponent`) but support multiple templates is really valuable. This is not an uncommon situation for our front end because we have a lot of "summary" areas with mostly the same data but very different display modes. This would be a way that ViewComponent could be better than how current front end component-based frameworks work.

For example, we're using a complex but mostly self-isolated Vue component as a spike test for ViewComponent, called `PricingSummary` (a high-level overview of how much a given project will cost to build). This component has three different distinct display variations (a horizontal, vertical, and dropdown), and they all use similar enough data that the same data object can represent all of them. Currently, we have to define three separate components in Vue with the same props and logic, which is a bummer.

Here's how we imagine this would look with ViewComponent:

```erb
<%= render(PricingSummaryComponent.new(template: 'vertical')) %>
```
```ruby
class PricingSummaryComponent < ApplicationComponent
  def initialize(template:)
    @template = template
  end

  # overrides parent method to specify a variant
  def template_variant
    # sample erb code above would load:
    #   pricing_summary_component.vertical.html.erb
    #
    # or if `template_variant` returns nil (the default), it would load
    #   pricing_summary_component.html.erb
    #
    # else if template isn't found, it's an error of course
    @template
  end
end
```

**First-class subfolder support**

Along with multiple template variants, our setup would greatly benefit from subfolder support. I understand that the GitHub use case doesn't currently require subfolders, but I'm certain this will be an issue for other people too.

I like the idea of following the order of how Rails looks for views, and first checking `app/components/component_name/component_name.rb` and then if that doesn't exist, checking `app/components/component_name.rb`. This would allow for both complex components with a bunch of support files, and simple self-rendering components with no support files.

With our test component, we already have a lot of support files, and I expect it to grow as third-party support of ViewComponent expands:

- We want a css file for each template variant
- We need a Stimulus controller for one variant.
- Maybe StimulusReflex can be awesomely loaded from  `app/components/component_name/component_name_reflex.rb` that sits along side your component instead of in `/app/reflexes`, etc.
- I see that the Storybook support (https://github.com/github/view_component/issues/159)
puts the stories in `test/components/stories/component_name_stories.rb`, but we'd prefer to keep them alongside our current component files (which is what we do in JS) if possible.

For our `PricingSummary` test component, this is the folder structure we'd ideally have today.

```
app/components/
  pricing_summary/
    pricing_summary_component.rb
    pricing_summary_component.dropdown.html.erb
    pricing_summary_component.dropdown.css
    pricing_summary_component.dropdown.js
    pricing_summary_component.horizontal.html.erb
    pricing_summary_component.horizontal.css
    pricing_summary_component.vertical.html.erb
    pricing_summary_component.vertical.css
    # Bonus third-party items
    pricing_summary_reflex.rb
    pricing_summary_component.stories.rb
```

**Reducing useless mini sub-components**

As @topherfangio [described above](https://github.com/github/view_component/issues/15#issuecomment-531281205), I think there's one feature of [`elemental_components`](https://github.com/jensljungblad/elemental_components) that is a big improvement over `view_component`, and that is specifying multiple bodies for content areas.

One of the annoying things about React and Vue is all of the "mini" components that you end up with that aren't actually reusable in any other context, like a `List` with a `ListItem`. The elemental\_components documentation does a great job of explaining this here: https://github.com/jensljungblad/elemental_components#elements

ViewComponent being able to specify multiple bodies with `component.with(:some_body)` in a similar way would be a big win. Otherwise we'll end up with a bunch of nonsense sub-component spaghetti code that is hard to maintain especially with Stimulus, which is based on top-level controllers rather than individual components having their own JavaScript.

Here's a simplified example of this use case:

```ruby
class MultiSelectListComponent < ApplicationComponent
  # add content area options
  with_content_areas :header, item: { multiple: true }
```
```erb
<!-- usage in a view -->
<%= render(MultiSelectListComponent.new) do |c| %>
  <% c.with(:header) do %>
    <i class="fas fa-temperature-high"></i>
    Permission List
  <% end %>

  <% permissions.each do |permission|
    <% c.with(:item) do %>
      <input type="checkbox" ...>

      <% if permission.readwrite? %>
        <i class="fas fa-lock-open"></i>
      <% else %>
        <i class="fas fa-lock"></i>
      <% end %>

      <%= permission_name %>
    <% end %>
  <% end %>
<% end %>
```
```erb
<!-- multiSelect_list_component.html.erb -->
<div data-controller="multiSelect-list">
  <div><%= header %></div>
  <ul>
    <!-- Proposal - with multiple bodies: AWESOME 🎉-->
    <% items.each_with_index do |item, index| %>
      <li class="multiSelect" data-target="multiSelect-list.item" data-index="<%= index %>">
        <%= item %>
      </li>
    <% end %>

    <!-- Current - without multiple bodies: YUCK 💩 -->
    <!-- requires a pointless separate component without multi-body support :( -->
    <%= render(MultiSelectListItemComponent.with_collection(items) %>
  </ul>
</div>
```

**Caching**

I believe this is covered in #234 but this is a really important one for us.

Thank you for reading and for this awesome gem! I am happy to go into more detail if something wasn't clear.

**Comment by jonspalmer** [comment]
@bbugh Great feedback 💯 Also mostly inline with my experience as we get deeper into using ViewComponents in anger. Hoping you can help explain your use cases. I'm confident we can address many of the ideas you bring up.

> Multiple views per component

I've faced similar situations. I switch between thinking the multiple views would be an AWESOME feature and considering it an anti-pattern better solved with composition of inheritance 😀 

That said I think we should consider it. It would be super helpful to have some concrete examples to look at and some sketches of what an ideal API would be to supply the variants. It's important to note that all the 'variations' of the views today get "compiled" into methods on the class so we'd likely need an API that statically defines the fixed list of views that were available as well as the api to choose them at render time.

> First-class Subfolder Support

This is subtle given how Ruby handles namespaces. You can make it work like this today by name spacing your components. In your example `class PricingSummary::PricingSummaryComponent` etc. It's not particularly pretty but it works.

I'd love there to be a solution like you propose - it would be very natural. I'm not an expert in how file loading/class resolution happens in Ruby/Rails. Clearly Rails is configured so that folders all subfolders of `app` don't get considered for class namespacing. Perhaps we can extend that somehow. I'd love to know how.

WRT Stories you can already configure your stories to be anywhere you like: https://github.com/jonspalmer/view_component_storybook#configuration
To put stories next to your component classes you should just be able to do 
```ruby
config.view_component_storybook.stories_path = Rails.root.join("app/components")
```

> Reducing useless mini sub-components

Multiple bodies for content-areas should be pretty achievable. I'd love your opinion on the API. Your suggestion of supplying it as part of `with_content_areas` would certainly work. I wonder if it's a bit more flexible to allow it in the actual calls to `with` that is:

```ruby
<% c.with(:item, append: true) %>
```

Reason being that I wonder if there are use cases where you want the ability to append and replace dynamically or is a given content_area always in one mode or the other?

I could also imagine we offer both - `with_content_areas` defines the default append/replace behavior and it can be overridden in calls the `with`. 

What do you think? I'd love more example use cases to validate the design against.

> Allowing component.with() to take options

This ask is captured in #285. Good timing because we have some attempts at providing this #323.
I'm concerned that some of the ways that this can be used would be anti patterns that passing things to the content areas would be better as properties of the component itself. It is rare that there are parameters only present in the components view that would _have_ to be passed in the view.

I'm a bit confused by your example NavComponent. 

1. It seems to assume that in the 'multiple' content areas feature you can iterate over the multiple content areas and template them. This might be cool feature but do we need it? If so why and can you share more examples? 

2. Why isn't the solution in your NavComponent to do this which doesn't require passing arguments:
```ruby
<%= render(NavComponent.new) do |c| %>
  <% pages.each do |page| %>
    <!-- need to pass in arguments here -->
    <% c.with(:item) do %>
      <li class="nav-item <%= 'active' if current_page?(page.path) %>">
        <%= page_icon(page) %>
        <%= page.name %>
      </li>
    <% end %>
  <% end %>
<% end %>
```
```ruby
<!-- nav_component.html.erb -->
... 
    <%= item %>
... 
```

I'd like to collect more use-cases for passing arguments to content areas. It seems like `form_with` is a good one. You can't do that any other way. Do you have more examples we can validate a design against?

**Comment by bbugh** [comment]
Hey, thanks for the reply! I appreciate how engaged y'all are in making this as awesome as possible. It's refreshing from some of the other open source projects I've worked with lately. 😀

> **Multiple views per component**
>
> I switch between thinking the multiple views would be an AWESOME feature and considering it an anti-pattern better solved with composition of inheritance 😀

I think you're right, maybe it can go either way. 

I'm not sure how often our specific use case of multiple templates with one view model would come up for other people. We were thinking of it in the way that [DHH's caching article](https://signalvnoise.com/posts/3112-how-basecamp-next-got-to-be-so-damn-fast-without-using-much-client-side-ui) describes multiple variations of their components (under "Thou shall share a cache between pages"). In their case it's one template modified by CSS, in our case it's multiple templates just because of differently shaped HTML requirements, but basically the same thing. E.g. the `PricingSummaryComponent` is a view model, with multiple views for the model. 

On the other hand, I think that separate component classes (even if they're a one line `include` is a better Ruby-friendly object-oriented design, and probably the best choice now that we've discussed it.

Given that Component object composition already supports this without new features and the existing solution is probably better than our proposal, this seems like something that could be waited on.

> **First-class Subfolder Support**
> 
> This is subtle given how Ruby handles namespaces. You can make it work like this today by name spacing your components. In your example class PricingSummary::PricingSummaryComponent etc. It's not particularly pretty but it works.

Ahh, that isn't great, but you're right given Rails' file loading that might be the best option. I think I remember seeing that someone else solved it this way already. Sorry for repeating this!

> **Reducing useless mini sub-components**

> I wonder if it's a bit more flexible to allow it in the actual calls to with that is:
> ```
> <% c.with(:items, append: true) %>
> ```

That's a cool idea, but I think it makes the internal implementation more complicated and won't work for the given example.

If you're saying `append` as in "append the content to an internal array", then component templates would have to treat every input value like an array, even if the component doesn't want to repeat the element, like a header:

```erb
<!-- 
Header shouldn't be appendable but we have to handle the append case
-->
<% Array.wrap(header).each do |header_content| %>
    <div><%= header %></div>
<% end %>
```

Or, if you're saying `append` as in "append the given inner content to an internal string buffer" then that won't work for the `MultiSelectListComponent` example, because the inner content may need to be wrapped in something by the component, and the component may need `Enumerable` methods like the example. The component consumer shouldn't need to write all of the wrapping code:

```erb
<!-- multi_select_list_component.html.erb -->
<% items.each_with_index do |item, index| %>
  <li class="multiSelect" data-target="multiSelect-list.item" data-index="<%= index %>">
    <%= item %>
  </li>
<% end %>
```

I personally think it's a better API to let the component define whether an input is a list or not, rather than letting the consumer pass in anything as a list. I think this would best be explicitly defined in the `with_content_areas` call. I explain more about this in the next answer.

> **Allowing component.with() to take options**
> 

> I'm a bit confused by your example NavComponent.

To be honest, I'm a bit confused by your questions, so I'll do my best to explain!

> It seems to assume that in the 'multiple' content areas feature you can iterate over the multiple content areas and template them. 

If I understand what you mean by this, yes, that is correct - this is based on the requirement that `with_content_areas` can define an input as a list instead of a single value, as previously discussed. There are many cases where on an internal component, you need to handle the loops yourself. See below -

> Why isn't the solution in your NavComponent to do this which doesn't require passing arguments:

That's an easy one to answer - because the consumer of the component shouldn't need to know the internals of how the component is rendered to use it! That's very error prone, hard to refactor/enhance and can get quite complicated when you get Stimulus involved. 

Let's make it real-world complicated so you can see why you wouldn't do this. Below is a reusable timeline component. We use something like this in multiple ways - a history of project updates, a list of recent user comments, a team's onboarding of new users, etc. Each of these is backed by a list of many different models or polymorphic lists, so we can't tie the TimelineComponent to any particular model.

Using your method without arguments, the TimelineComponent consumers would have to do this:

```erb
<%= render(TimelineComponent.new(option1: X, option2: Y) do |c| %>
  <% activities.each_with_index do |activity, index| %>
    <% c.with(:item) do %>
      <div class="timeline-item">
        <div class="timeline-icon">
          <%= fa_icon(activity.icon) %>
        </div>
        <div class="vertical-line"></div>
        <div class="activity-description" 
             data-target="timeline.item" 
             data-index="<%= index %>"
        >
          <%= activity.name %>- <%= time_ago_in_words(activity.created_at) %>
        </div>
      </div>
    <% end %>
  <% end %>
<% end %>
```

That unnecessarily exposes a TON about internals of the TimelineComponent to the consumer, when all the component needs beyond the `yield` input is the icon.

By allowing `with` to take in arguments, you get this instead:

```erb
<%= render(TimelineComponent.new) do |c| %>
  <% activities.each do |activity| %>
    <% c.with(:item, icon: activity.icon) do %>
      <%= activity.name %> - <%= time_ago_in_words(activity.created_at) %>
    <% end %>
  <% end %>
<% end %>
```

I hope that example speaks for itself! The consumer doesn't need to care at all about how the timeline renders, just what input it takes. 

Now imagine that this TimelineComponent came from a third party gem. Knowing all of the internals of how it works would be a mess! 

> It is rare that there are parameters only present in the components view that would have to be passed in the view.

If I understand what you mean by this, I would have to argue the other way, that it's *very* common, especially with reusable input-agnostic components.

In the above example, the `icon` is different for every item, so you can't just pass the icon in to `TimelineComponent`. You could also have an `icon_class` variable if you wanted to change the colors of the icons, for example. I could name a hundred of these example if the above sample code wasn't clear!

--- 

I hope that I answered everything! Happy to share more as needed.

**Comment by jonspalmer** [comment]
@bbugh Thank's for the details and patient response. I've broken the multiple content-area conversation into its own issue #325. Please keep the awesome feedback coming. You clearly have some deep practical experience with ViewComponents. Real-world problems and their solutions are crucial for us deliver the next set of features. Keep it coming. 🎉
